xdsl/irdl/attributes.py

```diff
@@ -280,7 +280,7 @@ def irdl_attr_definition(cls: TypeAttributeInvT) -> TypeAttributeInvT:
 
 IRDLGenericAttrConstraint: TypeAlias = (
     GenericAttrConstraint[AttributeInvT]
-    | Attribute
+    | AttributeInvT
```

Comment:
this definitely feels like a separate change to me, what's the advantage here?

Reply:
The issue is that without this change we have 
```
DenseArrayBase.constr(i64) : GenericAttrConstraint[DenseArrayBase[IntegerType | AnyFloat]]
```

Reply:
You can fix this by using `EqAttrConstraint` but it seems like an oversight in the definition of `IRDLGenericAttrConstraint` to me. I can move this to a separate change though I'm not sure how to test it.

Reply:
Sorry, I don't mean testing this, I mean testing the constr like the other tests in test_builtin. I think this would make sense in a separate PR.

Reply:
And the pyright bug seems fixed...

Reply:
> BTW I'm playing with this locally and it works just as well with `DenseArrayBase[I64]`

We want to move away from `Annotated` though?